Twig/NavigationExtension.php

```diff
@@ -2,63 +2,166 @@
 
 namespace Webfactory\Bundle\NavigationBundle\Twig;
 
+use Psr\Container\ContainerInterface;
+use Symfony\Contracts\Service\ServiceSubscriberInterface;
+use Twig\Environment;
 use Twig\Extension\AbstractExtension;
 use Twig\TwigFunction;
-use Webfactory\Bundle\NavigationBundle\Build\TreeFactory;
 use Webfactory\Bundle\NavigationBundle\Tree\Node;
 use Webfactory\Bundle\NavigationBundle\Tree\Tree;
 
-class NavigationExtension extends AbstractExtension
+class NavigationExtension extends AbstractExtension implements ServiceSubscriberInterface
 {
-    /** @var TreeFactory */
-    private $treeFactory;
+    /** @var ContainerInterface */
+    private $container;
 
-    public function __construct(TreeFactory $treeFactory)
+    public static function getSubscribedServices(): array
     {
-        $this->treeFactory = $treeFactory;
+        return [Tree::class];
+    }
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
     }
 
     public function getFunctions()
     {
         return [
+            new TwigFunction('navigation_tree', [$this, 'renderTree'], ['needs_environment' => true, 'is_safe' => ['html']]),
+            new TwigFunction('navigation_ancestry', [$this, 'renderAncestry'], ['needs_environment' => true, 'is_safe' => ['html']]),
+            new TwigFunction('navigation_breadcrumbs', [$this, 'renderBreadcrumbs'], ['needs_environment' => true, 'is_safe' => ['html']]),
             new TwigFunction('navigation_active_at_level', [$this, 'getNavigationActiveAtLevel']),
             new TwigFunction('navigation_find', [$this, 'findNode']),
             new TwigFunction('navigation_active_node', [$this, 'getActiveNode']),
             new TwigFunction('navigation_active_path', [$this, 'getActivePath']),
+            new TwigFunction('additional_navigation_item_classes', [$this, 'getAdditionalNavigationItemClasses']),
         ];
     }
 
     /**
-     * Returns the currently active tree node.
+     * Renders the navigation tree, starting at a given root node.
      *
-     * @return Node
+     * @param array|Node $root           Array of key-values pairs that will be passed to the
+     *                                   \Webfactory\Bundle\NavigationBundle\Tree\Tree::find method to look up the root node
+     * @param int        $maxLevels      Maximum number of tree levels (starting from the specified root) to draw
+     * @param int        $expandedLevels Number of levels to always draw expanded (i. e. showing all nodes).
      */
-    public function getActiveNode()
+    public function renderTree(
+        Environment $environment,
+        $root,
+        int $maxLevels = 1,
+        int $expandedLevels = 1,
+        string $template = 'WebfactoryNavigationBundle:Navigation:navigation.html.twig'
+    ): string {
+        $root = $this->getRootNode($root);
```

Comment:
Sollten wir hier nicht lieber sagen, `if (!$root instanceof Node) { $root = $this->findNode($root); }`?

Dann könnte die ganze `getRootNode()` Methode vielleicht entfallen und wir nutzen `findNode()`, die eh schon da ist?